src/cljam/algo/pileup.clj

```diff
@@ -142,6 +115,166 @@
+  in the corrected quals map."
+  [corrected-map ^SAMAlignment aln]
+  (let [qname (.qname aln)
+        tlen  (.tlen aln)]
+    (if (zero? tlen)
+      aln
+      (if-let [new-quals (get-in corrected-map
+                                 [(keyword qname) (keyword (str tlen))])]
+        (assoc aln :quals-at-ref new-quals)
+        aln))))
+
+(defn- get-max-pnext
+  "Get the maximum pnext of alignments."
+  [alns]
+  (->> alns
+       (map #(.pnext ^SAMAlignment %))
+       (apply max)))
+
+(defn- read-chunk
+  "Read a chunk which is specified by the region."
+  [sam-reader min-map-quality region]
+  (->> region
+       (sam/read-alignments sam-reader)
+       (sequence
+         (comp
+           (filter (basic-mpileup-pred min-map-quality))
+           (map index-cigar)))))
+
+(defn- get-chunk
+  "Get a chunk in which all pairs exist.
+  Returns the end position of the chunk and alignments in the chunk.
+  It concatenates chunks when a pair overlap across multiple chunks."
+  [sam-reader min-map-quality last-end {:keys [chr start end] :as region}]
+  (loop [s    start
+         e    end
+         alns (read-chunk sam-reader min-map-quality region)]
+    (if (or (empty? alns)                ;; empty chunk
+            (= e last-end)               ;; this is the last chunk
+            (<= (get-max-pnext alns) e)) ;; all pairs are in this chunk
+      [e alns]
+      (let [next-start (inc e)
+            next-end   (min last-end (+ e (- e s) 1))]
+        (recur
+          next-start
+          next-end
+          (->> (read-chunk sam-reader
+                           min-map-quality
+                           {:chr   chr
+                            :start next-start
+                            :end   next-end})
+               (concat alns)
+               distinct))))))
+
+(defn- pileup-chunk
+  "Piles up alignments in a chunk position-by-position.
+  Returns a lazy sequence of `cljam.io.pileup.LocusPile`s."
+  [^long start ^long end ignore-overlaps? alns]
+  (letfn [(step [^long pos prev-buf rest-alns]
+            (lazy-seq
+             (when (< (dec pos) end)
+               (let [[i rests] (split-with (partial cover-locus? pos) rest-alns)
+                     p (some-> rests ^SAMAlignment first .pos)
+                     [b :as buf] (into (filterv #(<= pos (.end ^SAMAlignment %))
+                                                prev-buf) i)
+                     next-pos (if (and p (not b)) p (inc pos))]
+                 (if b
+                   (cons [pos buf] (step next-pos buf rests))
+                   (when (and p (<= p end))
+                     (step next-pos buf rests)))))))]
+    (let [corrected-map (if ignore-overlaps?
+                          {}
+                          (make-corrected-quals-map alns))
+          [x :as xs] (drop-while #(< (.end ^SAMAlignment %) start) alns)]
+      (when x
+        (->> xs
+             (keep (partial correct-quals-at-ref corrected-map))
```

Comment:
It seems like that `pileup-chunk` and `pileup-seq` are very similar functions except the filtering behavior, so it might be a good way to define a more general function. For example, make the local function `step` as private function of the library, or define a function that takes same arguments (`start`, `end`, `alns`) and a new filter function (`#(keep (partial correct-quals-at-ref corrected-map) %)` for `pileup-chunk`, and `identity` for `pileup-seq`).

Reply:
Thanks for the proposal!
As you said, they are redundant.
I think that `pileup-chunk` can simply call `pileup-seq` after applying `correct-quals-at-ref`.